plutus-core/executables/pir/Main.hs

```diff
@@ -1,106 +1,288 @@
+compileToPlc optimise (PIR.Program _ pirT) = do
     plcTcConfig <- PLC.getDefTypeCheckConfig PIR.noProvenance
     let pirCtx = defaultCompilationCtx plcTcConfig
     runExcept $ flip runReaderT pirCtx $ runQuoteT $ PIR.compileTerm pirT
   where
-    set' :: Lens' (PIR.CompilationOpts a) b
-      -> (COpts -> b)
-      -> PIRCompilationCtx a
-      -> PIRCompilationCtx a
-    set' pirOpt opt = set (PIR.ccOpts . pirOpt) (opt opts)
-
     defaultCompilationCtx :: PLC.TypeCheckConfig PLC.DefaultUni PLC.DefaultFun
-      -> PIRCompilationCtx a
+      -> PIR.CompilationCtx PLC.DefaultUni PLC.DefaultFun a
     defaultCompilationCtx plcTcConfig =
-      PIR.toDefaultCompilationCtx plcTcConfig
-      & set' PIR.coOptimize                     cOptimize
-
-loadPirAndCompile :: COpts -> IO ()
-loadPirAndCompile copts = do
-    pirProg <- loadPir $ cIn copts
-    putStrLn "!!! Compiling"
-    case compile copts pirProg of
-        Left pirError -> error $ show pirError
-        Right _       -> putStrLn "!!! Compilation successful"
-
-loadPirAndAnalyse :: IOSpec -> IO ()
-loadPirAndAnalyse ioSpecs = do
+      PIR.toDefaultCompilationCtx plcTcConfig &
+         PIR.ccOpts . PIR.coOptimize .~ optimise
+
+compileToUplc :: Bool -> PlcProg () -> UplcProg ()
+compileToUplc optimise plcProg =
+    let plcCompilerOpts =
+            if optimise
+            then PLC.defaultCompilationOpts
+            else PLC.defaultCompilationOpts & PLC.coSimplifyOpts . UPLC.soMaxSimplifierIterations .~ 0
+    in flip runReader plcCompilerOpts $ runQuoteT $ PLC.compileProgram plcProg
+
+loadPirAndCompile :: CompileOptions -> IO ()
+loadPirAndCompile (CompileOptions language optimise test inp ifmt outp ofmt mode)  = do
+    pirProg <- getPirProgram ifmt inp :: IO (PirProg PLC.SrcSpan)
+    if test then putStrLn "!!! Compiling" else pure ()
+    -- Now compile to plc, maybe optimising
+    case compileToPlc optimise (() <$ pirProg) of
+      Left pirError -> error $ show pirError
+      Right plcTerm ->
+          let plcProg = PLC.Program () PLC.defaultVersion (() <$ plcTerm)
+          in case language of
+            PLC  -> if test
+                    then putStrLn "!!! Compilation successful"
+                    else writeProgram outp ofmt mode plcProg
+            UPLC -> do  -- compile the PLC to UPLC
+              let uplcProg = compileToUplc optimise plcProg
+              if test
+              then putStrLn "!!! Compilation successful"
+              else writeProgram outp ofmt mode uplcProg
+
+
+---------------- Optimisation ----------------
+
+doOptimisations :: PirTerm PLC.SrcSpan -> Either (PirError UnitProvenance) (PirTerm UnitProvenance)
+doOptimisations term = do
+  plcTcConfig <- PLC.getDefTypeCheckConfig PIR.noProvenance
+  let ctx = PIR.toDefaultCompilationCtx plcTcConfig
+      term' = (PIR.Original ()) <$ term
+  runExcept $ flip runReaderT ctx $ runQuoteT $ PIR.simplifyTerm term'
+
+-- | Run the PIR optimisations
```

Comment:
Not a big deal but maybe this can go in `Common.hs` so PIR and PLC can use the same function.